### PR TITLE
Add grammars module

### PR DESCRIPTION
**Pull request by lapp0**
Fixes https://github.com/outlines-dev/outlines/issues/543

Rendered docs: https://github.com/lapp0/outlines/blob/cfg-examples/docs/quickstart.md#built-in-grammars

Makes it easy to add grammars to the module, define a lark file in `outlines/lark`, reference it in `grammars.py`.

Work
- [x] allow grammar `Path` to be passed to `CFGFSM` and reference `outlines/lark` as lark importables
- [x] Introduce `grammars.py` and minimal example, `outlines.grammars.json`
- [x] test `outlines.grammars.json` with `CFGFSM`
- [x] Add to docs
- [x] Awaiting https://github.com/outlines-dev/outlines/pull/544 merge

Out of scope
- Fixing `escaped_string` per https://github.com/outlines-dev/outlines/issues/456
- Adding SQL, Python, etc grammars

### CODE REVIEW

**Comment by rlouf** [comment]
I think we could rename `outlines.lark` to `outlines.grammars`? Otherwise this requires user to know what `lark` is and add cognitive overhead.

**Comment by rlouf** [comment]
Now that #544 has been merged you will need to rebase your brain on `main` and fix the merge conflicts. It would also be nice to have some documentation :) This can go in the "Utiltities" section under "Docs".

File: outlines/fsm/fsm.py [hunk 1]
```diff
@@ -189,17 +191,28 @@ def copy(self) -> "RegexFSM":
 class CFGFSM(FSM):
     """FSM to generate text that is in the language of a context-free grammar."""
 
-    def __init__(self, cfg_string: str, tokenizer: "Tokenizer"):
-        self.cfg_string = cfg_string
+    def __init__(self, cfg: Union[str, Path], tokenizer: "Tokenizer"):
+        self.cfg = cfg
         self.tokenizer = tokenizer
 
+        if isinstance(cfg, Path):
```
Comment (rlouf) [inline outlines/fsm/fsm.py:198]: Could we load the grammars as strings instead of adding code that is not directly related to building the FSM here?

Comment (lapp0) [reply outlines/fsm/fsm.py]: Great idea, much cleaner. Done.

**Comment by rlouf** [comment]
Good to merge, thank you!
